README.md
# py-rcsb_utils_ccdc

RCSB Python wrapper for CCDC search and geometrical analysis tools.

## Introduction

Wrapper utilities for performing similarity and substructure searches on the CCDC database and
performing geometrical analysis of individual structure entries using the CCDC Python API.
This module has internal dependencies on the CCDC Python API and CCDC/CSD database.
These dependencies require licenses and separate installation that is described
with the CCDC documentation.  The 2021 version of the CCDC API is supported in Python 3.7.

### Installation

Download the library source software from the project repository and set the
enviroment corresponding to the CCDC installation in the example script `ccdc-api-env.sh`.
This script provides an example for a typical `pyenv` virtual configuration.

To install using pip:

```bash
pip install rcsb.utils.ccdc
```

Set the following environmental variables as required for you CCDC database
and Python API installation.

```bash
export CSDHOME=/Applications/CCDC/CSD_2021
# Locate your Python installation in which CCDC Python API is installed.
PYROOT=$PYENV_ROOT/versions/3.7.9
export CSD_PYTHON_ROOT_PATH=$PYROOT
#
# For Linux
export LD_LIBRARY_PATH=$PYROOT/lib:$PYROOT/lib/python3.7/site-packages/ccdc/_lib:$LD_LIBRARY_PATH
# For Macos
export DYLD_LIBRARY_PATH=$PYROOT/lib/python3.7/site-packages/ccdc/_lib
export DYLD_FRAMEWORK_PATH=$PYROOT/lib/python3.7/site-packages/ccdc/_lib
```

To install from the source repository:

```bash

git clone --recurse-submodules https://github.com/rcsb/py-rcsb_utils_ccdc.git

cd py-rcsb_utils_ccdc
pip install -r requirements
# edit and set the enviroment in the following bash script ...
. ccdc-api-env.sh

```

Optionally, run test suite (currently Python versions 3.7.9) using
[setuptools](https://setuptools.readthedocs.io/en/latest/) or
[tox](http://tox.readthedocs.io/en/latest/example/platform.html):

```bash
  # edit and set the enviroment in the following bash script ...
  . ccdc-api-env.sh
  pip install -r requirements.txt
  python setup.py test

or simply run:

 # edit and set the enviroment in the following bash script ...
  . ccdc-api-env.sh
  tox
```

Installation is via the program [pip](https://pypi.python.org/pypi/pip).  To run tests
from the source tree, the package must be installed in editable mode (i.e. -e):

```bash
pip install -e .
```

A CLI is provided to simplify access to the wrapper functions accross Python versions
not directly supported by the CCDC Python API.  A further multiprocessing wrapper is provided
which uses this CLI.  This intermediate approach was taken as the direct implementation of the python multiprocessing module atop the CSD API was not viable.

```bash
# edit and set the enviroment in the following bash script ...
. ccdc-api-env.sh
python CcdcSearchExec.py --help
   -or-
ccdc_search_cli --help

usage: CcdcSearchExec.py [-h] [--mol_list_path MOL_LIST_PATH]
                         [--result_path RESULT_PATH]
                         [--search_type SEARCH_TYPE]
                         [--start_record START_RECORD]
                         [--end_record END_RECORD] [--csdhome CSDHOME]
                         [--python_lib_path PYTHON_LIB_PATH]
                         [--python_version PYTHON_VERSION]

optional arguments:
  -h, --help            show this help message and exit
  --mol_list_path MOL_LIST_PATH
                        Molecule file list path
  --result_path RESULT_PATH
                        Molecule file list path
  --search_type SEARCH_TYPE
                        Search type (similarity|substructure)
  --start_record START_RECORD
                        Starting record
  --end_record END_RECORD
                        End record
  --csdhome CSDHOME     Path to the CSD release (path to CSD_202x)
  --python_lib_path PYTHON_LIB_PATH
                        Path to Python library
  --python_version PYTHON_VERSION
                        Python library version (default: 3.7)

```


HISTORY.txt
# File: HISTORY.txt
#
13-Dec-2020 - V0.11 Created module and ported code from py27 version
16-Dec-2020 - V0.12 Added CcdcSearch and CcdcSearchExec and associated tests
12-Jan-2021 - V0.13 Add geometry analysis module and assocated tests, update documentation
15-Jan-2021 - V0.14 Add multiprocessing wrapper for the CcdcSearchExec()
26-Jan-2021 - V0.15 Move IndexUtils() to rcsb.utils.io, extend index metadata
31-Jan-2021 - V0.16 Update add hasDisorder() status.
 2-Feb-2021 - V0.17 Add accession codes to generated mol/sdf and mol2 files
 6-Feb-2021 - V0.18 Update dependencies and push to PyPi


rcsb/utils/ccdc/CcdcSearchExec.py
##
# File: CcdcSearchExec.py
# Date: 15-Dec-2020  jdw
#
#  Execution wrapper  --  for CCDC search operations (wraps up the py37 environment)
#
#  Updates:
#   15-Jan-2021 jdw add option to export search hit list.
#
##
__docformat__ = "restructuredtext en"
__author__ = "John Westbrook"
__email__ = "jwest@rcsb.rutgers.edu"
__license__ = "Apache 2.0"

import argparse
import logging
import os
import sys

from rcsb.utils.io.MarshalUtil import MarshalUtil

HERE = os.path.abspath(os.path.dirname(__file__))
TOPDIR = os.path.dirname(os.path.dirname(os.path.dirname(HERE)))

logging.basicConfig(level=logging.INFO, format="%(asctime)s [%(levelname)s]-%(module)s.%(funcName)s: %(message)s")
logger = logging.getLogger()


def main():
    parser = argparse.ArgumentParser()
    #
    parser.add_argument("--mol_list_path", default=None, help="Molecule file list path")
    parser.add_argument("--result_path", default=None, help="Molecule file list path")
    parser.add_argument("--search_type", default=None, help="Search type (similarity|substructure)")
    parser.add_argument("--start_record", default=None, help="Starting record")
    parser.add_argument("--end_record", default=None, help="End record")
    parser.add_argument("--csdhome", default=None, help="Path to the CSD release (path to CSD_202x)")
    parser.add_argument("--python_lib_path", default=None, help="Path to Python library")
    parser.add_argument("--python_version", default=None, help="Python library version (default: 3.7)")
    parser.add_argument("--hit_list_path", default=None, help="Path to list of molecule identifers with search results")
    #
    args = parser.parse_args()
    #
    try:
        pyLib = args.python_lib_path if args.python_lib_path else os.path.join(os.environ["PYENV_ROOT"], "versions", "3.7.9", "lib")
        pyVer = args.python_version if args.python_version else "3.7"

        csdHome = args.csdhome
        molFilePath = args.mol_list_path
        resultPath = args.result_path
        searchType = args.search_type
        startRecord = args.start_record
        endRecord = args.end_record
        hitListPath = args.hit_list_path
    except Exception as e:
        logger.exception("Argument processing problem %s", str(e))
        parser.print_help(sys.stderr)
        exit(1)
    #
    try:
        os.environ["CSDHOME"] = csdHome
        os.environ["LD_LIBRARY_PATH"] = "%s:%s/python%s/site-packages/ccdc/_lib:$LD_LIBRARY_PATH" % (pyLib, pyLib, pyVer)
        os.environ["DYLD_LIBRARY_PATH"] = "%s/python%s/site-packages/ccdc/_lib" % (pyLib, pyVer)
        os.environ["DYLD_FRAMEWORK_PATH"] = "%s/python%s/site-packages/ccdc/_lib" % (pyLib, pyVer)

        logger.info("Using CSDHOME %s", os.environ["CSDHOME"])
        logger.info("Using DYLD_LIBRARY_PATH %s", os.environ["DYLD_LIBRARY_PATH"])
        logger.info("Using DYLD_FRAMEWORK_PATH %s", os.environ["DYLD_FRAMEWORK_PATH"])

        from rcsb.utils.ccdc.CcdcSearch import CcdcSearch  # pylint: disable=import-outside-toplevel

        ccdcS = CcdcSearch(verbose=True)
        pL = ccdcS.getList(molFilePath, startRecord=startRecord, endRecord=endRecord)
        logger.info("Search file %s record length %r", molFilePath, len(pL) if pL else [])
        #
        hitL = []
        for ii, queryTargetPath in enumerate(pL, 1):
            _, fn = os.path.split(queryTargetPath)
            queryTargetId, _ = os.path.splitext(fn)
            #
            logger.info("(%d/%d) Start search for %r %r", ii, len(pL), queryTargetId, queryTargetPath)
            numHits = ccdcS.search(queryTargetId, queryTargetPath, resultPath, searchType=searchType)
            if numHits:
                hitL.append(queryTargetId)
        logger.info("%d searches completed - matched %d", len(pL), len(hitL))
        if hitListPath:
            mU = MarshalUtil()
            ok = mU.doExport(hitListPath, hitL, fmt="list")
            logger.info("Wrote hit list (%r) to %s", ok, hitListPath)
    except Exception as e:
        logger.exception("Failing with %s", str(e))

    # ----------------------- - ----------------------- - ----------------------- - ----------------------- - ----------------------- -


if __name__ == "__main__":
    main()
